mdata/cwr.go

```diff
@@ -1,23 +1,83 @@
 package mdata
 
 import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
 	"time"
 
-	"github.com/grafana/metrictank/mdata/chunk"
+	"github.com/tinylib/msgp/msgp"
+
 	"github.com/raintank/schema"
 )
 
+type ChunkSaveCallback func()
+
 // ChunkWriteRequest is a request to write a chunk into a store
+//go:generate msgp
 type ChunkWriteRequest struct {
-	Metric    *AggMetric
+	Callback  ChunkSaveCallback `msg:"-"`
 	Key       schema.AMKey
-	Chunk     *chunk.Chunk
 	TTL       uint32
-	Span      uint32
+	T0        uint32
+	Data      []byte
 	Timestamp time.Time
 }
 
 // NewChunkWriteRequest creates a new ChunkWriteRequest
-func NewChunkWriteRequest(metric *AggMetric, key schema.AMKey, chunk *chunk.Chunk, ttl, span uint32, ts time.Time) ChunkWriteRequest {
-	return ChunkWriteRequest{metric, key, chunk, ttl, span, ts}
+func NewChunkWriteRequest(callback ChunkSaveCallback, key schema.AMKey, ttl, t0 uint32, data []byte, ts time.Time) ChunkWriteRequest {
+	return ChunkWriteRequest{callback, key, ttl, t0, data, ts}
+}
+
+// ArchiveRequest is used by the whisper importer utilities to transfer the data
```

Comment:
types shouldn't declare which callers use them. this is bound to get out of date or incomplete quickly. if you want to know where a certain type is used and with much more accuracy, there's plenty of tools (such as go guru or vscode's find references)

Reply:
if you want to document, better to document what the type is for. e.g. one could say it's a complete representation of a metric, both its timeseries data as well as its metadata.
the implications of how you could use it are then more obvious without tying us down.